src/consensus.rs

```diff
@@ -446,38 +449,54 @@ impl Consensus {
     }
 
     pub fn start(&mut self) -> anyhow::Result<()> {
-        let mut t = Instant::now();
-        let mut timeout = Duration::from_millis(self.config.tick_period_ms);
+        let tick_period = Duration::from_millis(self.config.tick_period_ms);
+
+        let mut previous_tick = Instant::now();
 
         loop {
+            let mut elapsed = previous_tick.elapsed();
+            let mut proposed = 0;
+
             if !self
                 .try_promote_learner()
-                .map_err(|err| anyhow!("Failed to promote learner: {}", err))?
+                .context("failed to promote learner")?
             {
-                // If learner promotion was proposed - do not add other proposals.
-                let have_changes = self.propose_updates(timeout)?;
-                if !have_changes {
-                    self.try_sync_local_state()?;
+                let mut timeout = tick_period.saturating_sub(elapsed);
+
+                if elapsed < tick_period && self.propose_updates(timeout)? {
+                    elapsed = previous_tick.elapsed();
+                    proposed += 1;
+
+                    timeout = Duration::from_millis(5);
```

Comment:
Where does this 5ms come from?

Reply:
My head? :D

The idea here is to "wait for the first operation available on the channel" and then "batch up to 128 of the *already queued* operations, as well as everything that *will be* queued shortly (next 5ms in this case)".

Reply:
My thinking was, would it possibly make more sense to base this on the tick rate? Would it possibly make more sense to use something like `self.config.tick_period_ms / 2`?

Reply:
Maybe, but, IMO, that's not all too important in this case. We're gonna loop back into this shortly anyway. If we timeout by some fraction of a tick, then we batch more, if we do fixed timeout, then we (theoretically, at least) lower the latency, because we do `on_ready` faster and send messages faster.